events: fix: shared object references

samhmills opened the pull request:
Fixes the issue: https://github.com/joyent/node/issues/25466

When emitting to multiple listeners for the same event type, the for
loop iteration over these listeners shares the same arguments. This
means (for objects), the copy is only shallow, and changes made by one
listener are overwritten on the object for the next listener.

For example, if we pass an object with arrays: as below

``` js
EventEmitter.emit("event", {
  data: [{
      users: [9237895]
  }]
});
```

And the first listener sets the users array as null like so:

``` js
EventEmitter.on('event', function(packet) {
    var data = packet.data[0];
    data.users = null;
});
```

The second event listener will ALWAYS show the array as null. This is
because only a shallow copy of the arguments is made and not a deep
copy.

This fixes that issue, by enforcing a deep copy and removing all
references from the new arguments array passed to each listener.

Link to code used to reproduce this issue:
https://ghostbin.com/paste/v4ova

Link to issue:
https://github.com/joyent/node/issues/25466

jasnell commented:
@Hunchmun ... you should be able to squash these commits down easily by doing a `git reset --soft HEAD~3` then redoing the final commit followed by a `git push -f` to force the update on your branch. Makes things a bit cleaner when landing.

Hunchmun commented:
@jasnell Thanks for your help, completed.

cjihrig commented on doc/api/events.markdown:
```diff
@@ -44,6 +44,11 @@ calls passing the same combination of `event` and `listener` will result in the
 
 Returns emitter, so calls can be chained.
 
+**Note:** Due to Pass by Reference, any objects passed to event listeners will be 
```
How about this?

> **Note:** Objects are passed by reference, meaning that the same object can be passed to multiple listeners. If a listener needs to modify a shared object, a copy of the object should be made.

cjihrig replied:
I wonder if this note should go on `emit()` instead of `addListener()`.

jasnell replied:
Likely. and +1 on the revision.

Commit:
docs: event emitter behavior notice

Updated documentation as per the issue below:
https://github.com/joyent/node/issues/25466

Event listeners can alter parts of the passed object, in some
circumstances the changes are passed to the next listeners
due to pass by reference. This is documentation of that behavior.

jasnell commented:
@Hunchmun ... did you mean to close this?

Hunchmun commented:
@jasnell nope, my apologies, this was as a result of resetting the branch, forgot to reopen, 

Thanks,

jasnell commented:
Thanks @Hunchmun! ... LGTM

jasnell commented:
Landed in a24db43
